### ISSUE TITLE
[R_package] regression model issues

### ISSUE THREAD

**Issue by amjiuzi** [Microsoft/LightGBM#935]
hi , @Laurae2 , need u again,
my raw data has 3 types: integer, factor , numeric
1. is it necessary to trans them into numeric like:
```r
fac1 <- as.numeric(fac1) - 1
integer1 <- as.numeric(integer1)
```
2. after transing into numeric, is it right on building lgbdataset like:
```r
dtrain <- lgb.Dataset(data=as.matrix(train1 %>% select(-sold_price)),
                      label=train1%>%select(sold_price),
                      is_sparse=FALSE,
                      free_raw_data = FALSE,
                      colnames = colnames(train1 %>% select(-sold_price)),
                      categorical_feature = c('month1','loc_clus1',
                                              'loc_clus4','carroomlevel1'))

dtest <- lgb.Dataset.create.valid(dataset=dtrain,
                                  data=as.matrix(valid1%>%select(-sold_price)),
                                  label=valid1%>%select(sold_price)) 
valids <- list(test=dtest,train=dtrain)
```
3. after preparation above, i run downstairs and got errors, what should i do on modeling successfully: 
 
```r
lgb1 <- lgb.train(data=dtrain,
                   valids = valids,
                   objective = 'regression',
                   metric='12',
                   min_data =1,
                   learning_rate=0.05,
                   nrounds = 10,
                   early_stopping_rounds = 5) #min_data=1

[LightGBM] [Info] Total Bins 2046 
[LightGBM] [Info] Number of data: 8001, number of used features: 17
Error in self$setinfo(names(p), p[[1]]) : 
  (list) object cannot be coerced to type 'double' 
```

**Comment by amjiuzi** [comment]
···r
  livingrooms bathrooms useArea floor KC_score buildingArea basementArea sold_price month1 loc_clus1 loc_clus4
1           3      2.25    1670     1        8         1240          430     545000      2         3        28
2           4      2.50    3300     2       10         3300            0     785000      1         3        47
3           3      3.25    3190     2        9         3190            0     765000      0         3        43
4           5      2.50    2900     2        9         2900            0     720000     10         3        19
5           5      2.75    2040     1        7         1200          840     449500      5         2        30
6           2      1.00     780     1        7          780            0     248500      4         1        29
  price_1mi carroomlevel1 existYears repaireYear last1price last3price     priceUp
1  326.3473             3         41          74   395.8084   309.4930  0.27889294
2  237.8788             4         31          84   260.8150   245.1194  0.06403276
3  239.8119             3          8         107   229.5918   240.6046 -0.04577136
4  248.2759             3         26          89   324.4275   275.0362  0.17958090
5  220.3431             3         46          69   225.2607   219.4549  0.02645524
6  318.5897             4         57          58   170.0000   178.1961 -0.04599487
```

**Comment by Laurae2** [comment]
Data must be a numeric matrix.

**Comment by guolinke** [comment]
@Laurae2 
can you convert other data types to numeric inside R package ?

**Comment by Laurae2** [comment]
@guolinke `lgb.prepare` should deal with most common cases, but the best method is to use the base package function `data.matrix` which will attempt to convert everything to a single numeric matrix.

**Comment by amjiuzi** [comment]
@Laurae2  as your advice, i changed 
```r 
data=data.matrix(train1 %>% select(-ids1,-sold_price))
```
so, final code as bellow, but it still doesn't works well
```r
library(lightgbm)
dt_all4 <- dt_all3 %>% 
  select(-sold_date) %>%
  mutate(priceUp=(last1price-last3price)/last3price)

table(sapply(dt_all4,class))
integer numeric 
     12       7

train_dt <- dt_all4 %>% filter(sold_price!=0)
test_dt <- dt_all4 %>% filter(sold_price==0)
indx1 <- sample(c(0,1),size = nrow(train_dt),replace = TRUE,prob = c(0.8,0.2))  
train1 <- train_dt %>% filter(indx1==0)
valid1 <- train_dt %>% filter(indx1==1)

dtrain <- lgb.Dataset(data=data.matrix(train1 %>% select(-ids1,-sold_price)),
                      label=train1%>%select(sold_price),
                      is_sparse=FALSE,
                      free_raw_data = FALSE,
                      colnames = colnames(train1 %>% select(-ids1,-sold_price)),
                      categorical_feature = c('month1','loc_clus1',
                                              'loc_clus4','carroomlevel1'))

dtest <- lgb.Dataset.create.valid(dataset=dtrain,
                                  data=data.matrix(valid1%>%select(-sold_price,-ids1)),
                                  label=valid1%>%select(sold_price))


valids <- list(test=dtest)

lgb1 <- lgb.train(data=dtrain,
                  valids = valids,
                  objective = 'regression',
                  metric='12',
                  min_data =1,
                  learning_rate=0.1,
                  nrounds = 100,
                  #nfold=5,
                  early_stopping_rounds = 10) 

Error in self$setinfo(names(p), p[[1]]) : 
  (list) object cannot be coerced to type 'double'
```

**Comment by amjiuzi** [comment]
数据文件
链接: https://pan.baidu.com/s/1b1PRIY 
 密码: 38n9

```r
dt_all3 <- read.csv('./dt_all31.csv',header = TRUE)
```

**Comment by guolinke** [comment]
@amjiuzi 
it is better to use english. Many maintainers of LightGBM are not Chinese.

**Comment by Laurae2** [comment]
@amjiuzi What is the output of the following?

```r
str(data.matrix(train1 %>% select(-ids1,-sold_price)))
str(data.matrix(valid1%>%select(-sold_price,-ids1)))
str(train1%>%select(sold_price))
str(valid1%>%select(sold_price))
```

By the way, what is this parameter?

```r
lgb1 <- lgb.train(data=dtrain,
                  valids = valids,
                  objective = 'regression',
                  metric='12', # metric twelve?
                  min_data =1,
                  learning_rate=0.1,
                  nrounds = 100,
                  #nfold=5,
                  early_stopping_rounds = 10) 
```

Also, remove categorical parameters for testing, along with all the (non useful for debugging) other parameters.

Something like this:

```r
library(dplyr)
library(lightgbm)

dt_all3 <- read.csv('./dt_all31.csv',header = TRUE)

dt_all4 <- dt_all3 %>% 
  select(-sold_date) %>%
  mutate(priceUp=(last1price-last3price)/last3price)

table(sapply(dt_all4,class))

train_dt <- dt_all4 %>% filter(sold_price!=0)
test_dt <- dt_all4 %>% filter(sold_price==0)
indx1 <- sample(c(0,1),size = nrow(train_dt),replace = TRUE,prob = c(0.8,0.2))  
train1 <- train_dt %>% filter(indx1==0)
valid1 <- train_dt %>% filter(indx1==1)

bla1 <- data.matrix(train1 %>% select(-ids1,-sold_price))
bla2 <- train1%>%select(sold_price)
dtrain <- lgb.Dataset(data=bla1,
                      label=bla2)

bla3 <- data.matrix(valid1%>%select(-sold_price,-ids1))
bla4 <- valid1%>%select(sold_price)
dtest <- lgb.Dataset.create.valid(dataset=dtrain,
                                  data=bla3,
                                  label=bla4)


valids <- list(test=dtest)

lgb1 <- lgb.train(data=dtrain,
                  valids = valids,
                  objective = 'regression',
                  metric='l2',
                  nrounds = 5,
                  early_stopping_rounds = 5) 
```

**Comment by amjiuzi** [comment]
@Laurae2 
- part 1 :
```r
str(data.matrix(train1 %>% select(-ids1,-sold_price)))
 num [1:8014, 1:17] 3 4 3 5 5 2 4 2 2 2 ...
 - attr(*, "dimnames")=List of 2
  ..$ : NULL
  ..$ : chr [1:17] "livingrooms" "bathrooms" "useArea" "floor" ...
> str(data.matrix(valid1%>%select(-sold_price,-ids1)))
 num [1:1986, 1:17] 2 3 4 2 4 2 5 5 4 4 ...
 - attr(*, "dimnames")=List of 2
  ..$ : NULL
  ..$ : chr [1:17] "livingrooms" "bathrooms" "useArea" "floor" ...
> str(train1%>%select(sold_price))
'data.frame':	8014 obs. of  1 variable:
 $ sold_price: num  545000 785000 765000 720000 449500 ...
> str(valid1%>%select(sold_price))
'data.frame':	1986 obs. of  1 variable:
 $ sold_price: num  235000 350000 220000 299000 883000 ...
```

- part 2 : 
metric = '12'  is from our demo in help page coming with objective = 'regression' in params,
may be  'rmse' or sm better evaluate rules ??

- part3 : 
after running yr adjusted codes, sames error reported as
```r
Error in self$setinfo(names(p), p[[1]]) : 
  (list) object cannot be coerced to type 'double'
```

**Comment by Laurae2** [comment]
1. you are using a data.frame as a label. Only numeric vectors are accepted for the labels, giving you the error (a data.frame is a list, a list cannot be converted to numeric as is)

2. Where? (link?)

3. See 1

**Comment by amjiuzi** [comment]
got it,
```r
metric='l2'  # not twelve,  -_-||
bla2 <- train1$sold_price
# not ( bla2 <- train1%>%select(sold_price) )
```
sorry for my low mistake, 3ks for your patience

**Comment by Laurae2** [comment]
@amjiuzi Is this solved?
